algebra2/gaussian_elimination/gaussian_elimination.h

```diff
@@ -0,0 +1,566 @@
+     * operation::add};*/
+    constexpr auto operator|(operation o1, operation o2) -> operation {
+        return static_cast<operation>(std::to_underlying(o1) |
+                                      std::to_underlying(o2));
+    }
+    /*you can check which operations are allowed e.g.
+    bool is_swap_allowed{operation::swap & swap_or_add != operation::none};*/
+    constexpr auto operator&(operation o1, operation o2) -> operation {
+        return static_cast<operation>(std::to_underlying(o1) &
+                                      std::to_underlying(o2));
+    }
+
+    enum class axis {
+        rows,
+        columns,
+    };
+
+    /*If you perform (standard) gaussian elimination
+    you obtain echelon form (see
+    https://en.wikipedia.org/wiki/Gaussian_elimination#Echelon_form) Sometimes
+    (e.g. when you look for an inverse) you need a diagonal form.*/
+    enum class reducted_form {
+        echelon,
+        diagonal /*with ones or zeros on the diagonal and 0 otherwise*/
+    };
+    /*runs gaussian elimination on axis a for matrix m with allowed
+    operations given by allowed_operations until the reducted form reducted is
+    obtained*/
+    template <std::integral T, std::integral LP>
+    auto run(ranges::matrix_view<T, LP> m,
+             axis a,
+             operation allowed_operations = operation::swap | operation::add |
+                                            operation::multiply,
+             reducted_form reducted = reducted_form::echelon) -> void;
+
+
+    /*runs gaussian elimination on axis a for matrix m with allowed
+operations given by allowed_operations*/
+    /*Demonstrate on screen intermediate steps of the elimination*/
+    template<std::integral T, typename LP>
+    auto show_steps(ranges::matrix_view<T, LP> m,
+                    axis a,
+                    operation allowed_operations = operation::swap | operation::add | operation::multiply,
+                    reducted_form reducted = reducted_form::echelon) -> void;
+
+    enum class error {
+        not_invertible,
+        not_square,
+    };
+
+    // returns inverse of matrix m
+    template <typename T, typename LP>
+    auto inverse(ranges::matrix_view<T, LP> m) -> std::expected<ranges::matrix_view<T, LP>, algorithms::gaussian_elimination::error>;
+
+    /*calculates the determinant of matrix m*/
+    template <typename T, typename LP>
+    auto determinant(ranges::matrix_view<const T, LP> m) -> std::expected<T, error>;
+
+    /*calculates the determinant of matrix m containing fractions*/
+    template <std::integral T, typename LP>
+    auto determinant(ranges::matrix_view<ranges::fraction<T>, LP> m) -> std::expected<ranges::fraction<T>, algorithms::gaussian_elimination::error>;
+
+}  // namespace algorithms::gaussian_elimination
+
+
+/*description:
+ *      Performs gaussian elimination on matrix m containing fractions*/
+template<std::integral T, typename LP>
```

Comment:
wrap your implementations into namespace algorithms::gaussian_elimination {...} (otherwise you code repeats algorithms::gaussian_elimination::, algorithms::gaussian_elimination:: ...